sqlitecluster/SQLite.cpp

```diff
@@ -256,79 +308,139 @@ bool SQLite::write(const string& query) {
+    string query = "INSERT INTO " + _journalName + " VALUES (" + SQ(commitCount + 1) + ", " + SQ(_uncommittedQuery) + ", " + SQ(_uncommittedHash) + " )";
+
+    // These are the values we're currently operating on, until we either commit or rollback.
+    _uncommittedID = commitCount + 1;
+    _inFlightTransactions[commitCount + 1] = make_pair(_uncommittedQuery, _uncommittedHash);
+
+    int result = SQuery(_db, "updating journal", query);
     _prepareElapsed += STimeNow() - before;
-    if (!result) {
+    if (result) {
         // Couldn't insert into the journal; roll back the original commit
-        SWARN("Unable to prepare transaction, rolling back: " << _uncommittedQuery);
+        SWARN("Unable to prepare transaction, got result: " << result << ". Rolling back: " << _uncommittedQuery);
         rollback();
+        commitLock.unlock();
         return false;
     }
 
     // Ready to commit
     SDEBUG("Prepared transaction");
+
+    // We're still holding commitLock now, and will until the commit is complete.
     return true;
 }
 
-// --------------------------------------------------------------------------
-void SQLite::commit() {
+int SQLite::commit() {
     SASSERT(_insideTransaction);
     SASSERT(!_uncommittedHash.empty()); // Must prepare first
+
+    int result = 0;
     // Do we need to truncate as we go?
     bool truncating = false;
     if (_journalSize + 1 > _maxJournalSize) {
         // Delete the oldest entry
         truncating = true;
         uint64_t before = STimeNow();
-        SASSERT(SQuery(_db, "Deleting oldest row",
-                       "DELETE FROM journal WHERE id=" + SToStr(_commitCount - _journalSize + 1)));
+        string query = "DELETE FROM" + _journalName + " WHERE id = MIN(id);";
+        SASSERT(!SQuery(_db, "Deleting oldest row", query));
         _writeElapsed += STimeNow() - before;
     }
 
     // Make sure one is ready to commit
     SDEBUG("Committing transaction");
     uint64_t before = STimeNow();
-    SASSERT(SQuery(_db, "committing db transaction", "COMMIT"));
-    _commitElapsed += STimeNow() - before;
+    result = SQuery(_db, "committing db transaction", "COMMIT");
 
-    // Successful commit
-    ++_commitCount;
-    _journalSize += !truncating; // Only increase if we didn't truncate by a row
-    _committedHash = _uncommittedHash;
-    _insideTransaction = false;
-    _uncommittedHash.clear();
-    _uncommittedQuery.clear();
+    // If there were conflicting commits, will return SQLITE_BUSY_SNAPSHOT
+    SASSERT(result == SQLITE_OK || result == SQLITE_BUSY_SNAPSHOT);
+
+    if (result == SQLITE_OK) {
+        _commitElapsed += STimeNow() - before;
+        // Successful commit
+        _journalSize += !truncating; // Only increase if we didn't truncate by a row
+
+        _committedtransactionIDs.insert(_uncommittedID);
+        {
+            // Update atomically.
+            lock_guard<recursive_mutex> lock(_hashLock);
+            _lastCommittedHash = _uncommittedHash;
```

Comment:
Why not just define it as `atomic<string> _uncommittedHash` to avoid the extra mutex?